### ISSUE TITLE
Security - Can't use CSP that blocks 'unsafe-inline';

### ISSUE THREAD

**Issue by Graham42** [remix-run/remix#183]
To improve security and mitigate cross site scripting threats, sites should set a [Content Security Policy (CSP)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP). I would like to restrict to `script-src: self` to block inline scripts, but Remix crashes on hydration when this is set.

Remix version: 0.17.2

Workaround: don't hydrate, server rendered pages still work 👌 🚀

**Comment by dchristensen** [comment]
I'll add that the Scroll Restoration code also adds some additional script inline that would need to mitigated as well:
https://github.com/remix-run/remix/blob/main/packages/remix-react/scroll-restoration.tsx

**Comment by Graham42** [comment]
That blog post above that I wrote is out of date, and my memory of this topic has turned a bit fuzzy, but I think there should be a way in Remix to provide all the correct hashes for inline scripts, since Remix takes care of the bundling. I'm wary of the nonce approach as I remember that has been misused by other bundlers in the past. It's at odds with the desire to cache things as much as possible since the nonce value  must be a new unique random generated value for every page load. 
Things have been pretty busy for me so I haven't had a chance to look at this, but I'm hoping to revisit mid December.

**Comment by renchap** [comment]
Using hashes for inline scripts is probably a better idea than nonces, but this can be an issue for streaming responses as you need to send the headers before starting generating the content.

Does Remix start streaming the response before it has been fully computed?

**Comment by fnimick** [comment]
Any word on this? I'd like to set a content-security-policy for a new project and this is preventing us from doing so.

**Comment by ngbrown** [comment]
It's possible to use a `content-security-policy` with `nonce`. I could only use a `nonce` on the default export of `root.tsx`, not the `CatchBoundary` or `ErrorBoundary` exports.

```tsx
export const loader: LoaderFunction = async () => {
  const cspScriptNonce = cryptoGetRandomString(33);
  const data: RootLoaderData = {
    cspScriptNonce,
  };
  return data;
};

export const unstable_shouldReload = () => false;

export default function App() {
  const { cspScriptNonce } = useLoaderData<RootLoaderData>();
  return (
    <html lang="en">
      <head>
        <Meta />
        <Links />
      </head>
      <body>
        <Outlet />
        <ScrollRestoration nonce={cspScriptNonce} />
        <Scripts nonce={cspScriptNonce} />
        <LiveReload nonce={cspScriptNonce} port={8002} />
      </body>
    </html>
  );
}
```

And in `entry.server.tsx`, I added this to `handleRequest()`:

```tsx
    const nonce =
      remixContext.appState.catchBoundaryRouteId === "root" &&
      remixContext.appState.error
        ? undefined
        : remixContext.routeData.root?.cspScriptNonce;
    responseHeaders.set(
      "Content-Security-Policy",
      getContentSecurityPolicy(nonce)
    );
```

Most of the complexity is in dynamically generating `getContentSecurityPolicy()` based on the presence of a `nonce` and `NODE_ENV`:

```tsx
function getContentSecurityPolicy(nonce?: string) {
  let script_src: string;
  if (typeof nonce === "string" && nonce.length > 40) {
    script_src = `'self' 'report-sample' 'nonce-${nonce}'`;
  } else if (process.env.NODE_ENV === "development") {
    // Allow the <LiveReload /> component to load without a nonce in the error pages
    script_src = "'self' 'report-sample' 'unsafe-inline'";
  } else {
    script_src = "'self' 'report-sample'";
  }

  const connect_src =
    process.env.NODE_ENV === "development"
      ? "'self' ws://localhost:*"
      : "'self'";

  return (
    "default-src 'self'; " +
    `script-src ${script_src}; ` +
    "style-src 'self' 'report-sample'; img-src 'self' data:; font-src 'self'; " +
    `connect-src ${connect_src}; ` +
    "media-src 'self'; object-src 'none'; " +
    "prefetch-src 'self'; " +
    "child-src 'self'; " +
    "frame-src 'self'; worker-src 'self' blob:; frame-ancestors 'none'; " +
    "form-action 'self'; " +
    "block-all-mixed-content; " +
    "base-uri 'self'; manifest-src 'self'"
  );
}
```

Also, `react-dom` generates some error messages because it doesn't handle rehydration checking correctly on `nonce` since it can't be read as an attribute.

**Comment by davidpfahler** [comment]
@ngbrown Is there anything we can do about the "Prop `nonce` did not match." warning? Why is there a mismatch even? Why is the nonce empty string on the server?

**Comment by ngbrown** [comment]
Hi @davidpfahler, I'm sure you're discovering that CSP is a fairly complex feature. With as little information you've given it's hard to know what is going on. Maybe open a discussion for more interactive follow up?

**Comment by davidpfahler** [comment]
@ngbrown Sorry for being unclear. I was referring to your earlier comment saying

> Also, react-dom generates some error messages because it doesn't handle rehydration checking correctly on nonce since it can't be read as an attribute.

I thought that what you meant by that was that there is an error in the browser console, namely:

```
Warning: Prop `nonce` did not match. Server: "" Client: "irn2qSvqJ-0xG9HuIos8DrJjQBtqYMk8L3PVz4qyG74muQ"
```

**Comment by EvgeniyBudaev** [comment]
@ngbrown  Where can I get responseHeaders? Where can i see code cryptoGetRandomString?

**Comment by ngbrown** [comment]
`responseHeaders` is a parameter of `handleRequest()` in your `entry.server.ts` file. `cryptoGetRandomString()` is an exercise left to the reader as it differs per platform :). The `33` is just the number of bytes of randomness I wanted.

**Comment by EvgeniyBudaev** [comment]
@ngbrown I get a warning too in console Warning: Prop `nonce` did not match. Server: "" Client: "irn2qSvqJ-0xG9HuIos8DrJjQBtqYMk8L3PVz4qyG74muQ"  I really need help on how to fix this

**Comment by ngbrown** [comment]
This is probably the warning from React that I was refering to:

> `react-dom` generates some error messages because it doesn't handle rehydration checking correctly on `nonce` since it can't be read as an attribute.

I used `patch-package` to fix this, or you can just ignore it if your CSP is working correctly.

**Comment by EvgeniyBudaev** [comment]
To fix "Warning: Prop nonce did not match. Server: "" Client: "irn2qSvqJ-0xG9HuIos8DrJjQBtqYMk8L3PVz4qyG74muQ" ", you need to add:
`
	if(typeof window !== "undefined") {
		cspScriptNonce = "";
	}
`
Full code:
`
export default function App() {
	const { cspScriptNonce } = useLoaderData<RootLoaderData>();
	if(typeof window !== "undefined") {
		cspScriptNonce = "";
	}
	return (
	  <html lang="en">
		<head>
		  <Meta />
		  <Links />
		</head>
		<body>
		  <Outlet />
		  <ScrollRestoration nonce={cspScriptNonce} />
		  <Scripts nonce={cspScriptNonce} />
		  <LiveReload nonce={cspScriptNonce} port={8002} />
		</body>
	  </html>
	);
  }
`

**Comment by fnimick** [comment]
@EvgeniyBudaev thank you! This works great!

For anyone unclear on why this works - the nonce attribute is stripped from the script elements as soon as the page loads. Then when client hydration happens, the nonce values are re-applied, and are then updated again every time the loader is re-fetched. This shouldn't be happening anyway, and it would be great if there was added documentation in the Remix docs to ensure the nonces are only provided on the initial server-side load.

ref: https://html.spec.whatwg.org/multipage/urls-and-fetching.html#nonce-attributes : "Elements that have a [nonce](https://html.spec.whatwg.org/multipage/urls-and-fetching.html#attr-nonce) content attribute ensure that the cryptographic nonce is only exposed to script (and not to side-channels like CSS attribute selectors) by taking the value from the content attribute, moving it into an internal slot named [[CryptographicNonce]], exposing it to script via the [HTMLOrSVGElement](https://html.spec.whatwg.org/multipage/dom.html#htmlorsvgelement) interface mixin, and setting the content attribute to the empty string. Unless otherwise specified, the slot's value is the empty string."

**Comment by 58bits** [comment]
Was also looking at this today. Not sure if this helps - but Kent Dodd's site @kentcdodds  - is also using nonce values, generated in his express server, and available via loader context.....  https://github.com/kentcdodds/kentcdodds.com/blob/main/app/root.tsx

**Comment by ShawLocke** [comment]
@ngbrown thanks for your comment. i am inspired by it and setting CSP for shopify app with remix successfully which is critical for shopify app review and routine evaluation.

**Comment by brophdawg11** [comment]
I think this is no longer an issue with the `nonce` support?  As shown on Kent's site - the nonce can be included on the `<script>` tags rendered via `<Scripts>`, `<LiveReload>`, `<ScrollRestoration>` etc., and then included in the CSP accordingly.

If this is still an issue, please feel free to re-open with a small reproduction.

**Comment by kentcdodds** [comment]
This is correct. It has been fixed 👍

**Comment by samcolby** [comment]
For anyone as confused by this thread as I was... At the time of writing this isn't currently working on [Kent's site ](https://github.com/kentcdodds/kentcdodds.com/blob/main/server/index.ts#L252). 

And the [NonceContext](https://github.com/remix-run/remix/issues/5162) approach doesn't add the nonce to the html, which I assume is why it is not working on Kent's site.

Using the other comments in this thread, you can get it work using remix `1.19.3` with all V2 flags enabled using 

In `root.tsx`
```
export const loader: LoaderFunction = async () => {
  return {cspScriptNonce: crypto.randomBytes(16).toString('hex')}
}

// don't include any inline scripts with whatever you use here as the nonce wont work
export const ErrorBoundary = () => <div /> 

const App = () => {
    const data = useLoaderData<LoaderData>() 
    const cspScriptNonce = typeof document === 'undefined' ? data.cspScriptNonce : ''
    ...etc...
}
```

Then in `handleRequest` or `handleBrowserRequest` etc..
```
const cspScriptNonce = remixContext.staticHandlerContext.loaderData.root?.cspScriptNonce
// this needs to handle cspScriptNonce=undefined for ErrorBoundaries
applySecurityHeaders(responseHeaders, cspScriptNonce). ```
```

Love to know if there is a better way of doing this without using the `loader` though.  Or ideally, we don't have any inline scripts to worry about.

**Comment by ahuth** [comment]
Hmm, [seems to be working for me with v1.19.3](https://github.com/ahuth/andrew-stack/blob/6470a1c9bd7fa63fd2d8b11c9d79b1d78cc3cb85/server.ts#L50-L54)

**Comment by samcolby** [comment]
Aha!  Thank you @ahuth that's a much better easier to follow!

**Comment by Haraldson** [comment]
@samcolby @ahuth How are you piecing this together? Adding the nonce itself doesn’t seem to be enough. Is there a guide somewhere?

I’ve added …

#### `server.mjs`

```javascript
import express from 'express'
import helmet from 'helmet'

const server = express()

server.use((request, response, next) => {
    response.locals.cspNonce = crypto.randomBytes(32).toString('base64')
    next()
})

// ...

server.use(helmet({
    contentSecurityPolicy: {
        directives: {
            'script-src': [
                "'strict-dynamic'",
                (request, response) => `'nonce-${response.locals.cspNonce}'`
            ]
        }
    }
}))

// ...

server.all('*', createRequestHandler({
    build,

    getLoadContext: (request, response) => ({
        cspNonce: response.locals.cspNonce
    })
}))
```

#### `entry.server.js` (both bot and browser handlers)

```javascript
import NonceProvider from '@providers/nonce'

// ...

return new Promise((resolve, reject) => {
    let shellRendered = false

    const nonce = loadContext?.cspNonce

    const {
        pipe,
        abort
    } = renderToPipeableStream(
        <NonceProvider nonce={nonce}>
            <RemixServer
                context={remixContext}
                url={request.url}
                abortDelay={ABORT_DELAY} />
        </NonceProvider>,
        {
            nonce, // <-- Unsure about this one
            
            // ...
```

#### `providers/nonce.js`

```javascript
import { createContext, useContext } from 'react'
const NonceContext = createContext('')
export const useNonce = () => useContext(NonceContext)
export default NonceProvider = NonceContext.Provider
```

#### `root.js`

```javascript
import { useNonce } from '@providers/nonce'

// ...

const App = () => {
    const nonce = useNonce()

    return (
        <html>
            <head>
                <Meta />
                <Links />
            </head>
            <body>
                <Outlet />
                <Scripts nonce={nonce} />
                <ScrollRestoration nonce={nonce} />
                <LiveReload nonce={nonce} />
            </body>
        </html>
    )
}


… and still CSP is whining in my console.

Did I forget anything obvious?

**Comment by cskeppstedt** [comment]
> ```js
> getLoadContext
> ```

I'm trying to implement something similar, but I'm stuck on getting the object with `{cspNonce: '...'}` from `getLoadContext()` passed to `handleRequest()` in `entry.server.tsx` – the last argument (loadContext) is always undefined.

I can see the console.log working in my express server:

```js
app.all(
  '/admin/*',
  createRequestHandler({
    build,
    getLoadContext: (request, response) => {
      console.log('getLoadContext: hello', response.locals); // <-- this works: getLoadContext: hello, {cspNonce: 'foobar123....;}
      return { cspNonce: response.locals.cspNonce };
    },
  }),
);
```

but the loadContext argument is always undefined:

```js
export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext,
  loadContext?: AppLoadContext,
) {
  console.log(loadContext) // <-- always undefined
  return ...
}
```

I also tried to look for the `cspNonce` value inside `remixConfig` but could not find it anywhere (some code example suggested `remixContext.staticHandlerContext.loaderData.root.cspNonce`, but I also console logged all arguments and it was nowhere to be found).

I'm on remix `v1.19.3` (latest 1.x release).

**Comment by outofthisworld** [comment]
How come <Links/> does not support nonce?

**Comment by brophdawg11** [comment]
Would you mind starting a new issue for Links nonce support @outofthisworld?  From a quick glance it does seem that we should support a nonce to be applied to link requests: https://html.spec.whatwg.org/#fetching-and-processing-a-resource-from-a-link-element

**Comment by lilouartz** [comment]
How do I add nonce to inline styles?

**Comment by saaskitten** [comment]
> How do I add nonce to inline styles?

I worked around as followed.

```
const LinksWithNonce = ({ nonce }: { nonce: string }) => {
  return (
    <Fragment>
      {Links().props.children[1].map((child: JSX.Element) =>
        cloneElement(child, { nonce }),
      )}
    </Fragment>
  );
};

return (
  <html>
    <head>
       <LinksWithNonce nonce={nonce.style} />
    </head>
  </html>
)
```

**Comment by cha0s** [comment]
Just wanted to make a note that I'm still seeing breaks with the inline script that looks like  `$RC("B:1","S:1")` as well as the `window.__remixContext.streamController.*` inline scripts.

My current solution is using a transform stream on the `createReadableStreamFromReadable` call in `entry.server.jsx` to transform all my script tags. Something like:

```
  class Noncer extends Transform {
    constructor(nonce) {
      super();
      this.nonce = nonce;
    }
    _transform(chunk, encoding, done) {
      const string = chunk.toString();
      this.push(string.replaceAll('<script', `<script nonce="${this.nonce}"`));
      done();
    }
  }
```

Note that this is a node-specific thing. It would obviously be better for e.g. the `<Scripts>` tag to respect the nonce for any generated script.

**Comment by einarq** [comment]
@brophdawg11 
Any update on supporting a nonce on the streamController stuff as well? When setting a nonce on the Scripts element we still get this script. 

`<script>window.__reactRouterContext.streamController.enqueue`

Should it be a new ticket?

**Comment by einarq** [comment]
@cha0s 

FYI: https://github.com/remix-run/remix/issues/5162#issuecomment-2749116785

**Comment by einarq** [comment]
So I got a bit further, but RR7 still seems to generate some script tags that don't get the nonce. They seem related to retries of the streaming? Not sure.

`<script>function $RC(a,b){a=document.getElementById(a);b=document.getElementById(b);b.parentNode.removeChild(b);if(a){a=a.previousSibling;var f=a.parentNode,c=a.nextSibling,e=0;do{if(c&&8===c.nodeType){var d=c.data;if("/$"===d)if(0===e)break;else e--;else"$"!==d&&"$?"!==d&&"$!"!==d||e++}d=c.nextSibling;f.removeChild(c);c=d}while(c);for(;b.firstChild;)f.insertBefore(b.firstChild,c);a.data="$";a._reactRetry&&a._reactRetry()}};$RC("B:0","S:0")</script>`

`<script>$RC("B:1","S:1")</script>`

`<script>$RC("B:2","S:2")</script>`

These seem static tho, so for now I'm rolling with the sha that Chrome generated for me.

**Comment by brophdawg11** [comment]
There are a handful of spots that require `nonce` - please confirm each of these have them and if they do and this is still an issue please file a new issue in the react router repo for a fix in v7

 - `root.tsx` - `<Scripts nonce>`
 - `entry.server.tsx` - `<ServerRouter nonce>`
 - `entry.server.tsx` - `renderToPipeableStream(..., { nonce })`

**Comment by einarq** [comment]
@brophdawg11 
Thanks! I was missing the one on renderToPipeableStream. After adding that the nonce is finally on all the script tags that gets produced.

Its also needed on ScrollRestoration btw.

**Comment by einarq** [comment]
Which means that @kentcdodds was probably also missing that in the Epic Stack example I think.

**Comment by brophdawg11** [comment]
yeah we should add docs on that - opened a PR to get that started in <https://github.com/remix-run/react-router/pull/13356>

**Comment by priteshloke** [comment]
const { pipe, abort } = renderToPipeableStream(
      serverElement,
      {
        nonce, // ✅ this makes React attach nonce to its internal <script> tags
        [readyOption]() {
          shellRendered = true;
          const body = new PassThrough();
          const stream = createReadableStreamFromReadable(body);

          responseHeaders.set("Content-Type", "text/html");
          responseHeaders.set(
            "Content-Security-Policy",
            `default-src 'self'; script-src 'self' 'nonce-${nonce}'; style-src 'self' 'nonce-${nonce}' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; object-src 'none'; upgrade-insecure-requests`
          );

          resolve(
            new Response(stream, {
              headers: responseHeaders,
              status: responseStatusCode,
            })
          );

          pipe(body);
        },
        onShellError(error: unknown) {
          reject(error);
        },
        onError(error: unknown) {
          responseStatusCode = 500;
          // Log streaming rendering errors from inside the shell.  Don't log
          // errors encountered during initial shell rendering since they'll
          // reject and get logged in handleDocumentRequest.
          if (shellRendered) {
            console.error(error);
          }
        },
      }
    );

added in renderToPipeableStream but still not coming in <script>window.__reactRouterContext.streamController

**Comment by brophdawg11** [comment]
Please open a new issue with a reproduction on the react router repo
